zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,263 @@
+- ZAP PR: [zulip/zulip-architecture#2](https://github.com/zulip/zulip-architecture/pull/2)
+
+# Summary
+
+Proposal to add end-to-end encryption for mobile push notifications.
+
+# Motivation
+
+Currently, the Zulip Push Notification Forwarding Service receives
+(but does not store or log) the contents of individual mobile push notifications.
+
+Adding end-to-end encryption for push notifications will ensure that the
+actual message content does not need to be readable by the push notification
+bouncer (or Google/Apple's servers, for that matter).
+
+# Detailed design
+
+## High-Level Plan
+
+- Client checks the server if it supports E2E encryption. If it doesn't,
+  use the legacy push notification registration flow.
+
+- Client generates an asymmetric key pair (public and private key).
+  It stores the private key securely and makes an API call to the server
+  with the public key to register itself for push notifications.
+
+- While sending notification, the server encrypts the notification content
+  to the public key from above and sends it to push bouncer.
+  Push bouncer forwards it to FCM/APNs.
+
+- Client receives the notification via FCM/APNs and decrypts the content
+  using the private key it stored earlier.
+
+## Data model
+
+- On the client, include the following fields on [the Account type](https://github.com/zulip/zulip-flutter/blob/ec9aa351b3050a0fa3e467007826576110fdbf99/lib/model/database.dart#L66-L94):
+  ```
+  model Account:
+    …
+    push_account_id: int  # 64-bit random integer generated by client; unique here.
+    last_acked_push_token: str  # The last push token the server acknowledged receipt of from this client.
+    push_private_key: PrivateKey  # Private part of the asymmetric key pair generated by the client.
```

Comment:
Noticed this in the context of https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057020659 :
```suggestion
    push_private_key: PrivateKey  # Private part of the asymmetric key pair generated by the client.
    push_public_key: PublicKey  # Public key corresponding to push_private_key
```

The client needs to remember the public key at least long enough to retry the registration request.